pkg/build/builder/daemonless.go

```diff
@@ -0,0 +1,293 @@
+}
+
+func buildDaemonlessImage(dir string, optimization buildapiv1.ImageOptimizationPolicy, opts *docker.BuildImageOptions) error {
+	glog.V(0).Infof("Building...")
+
+	args := make(map[string]string)
+	for _, ev := range opts.BuildArgs {
+		args[ev.Name] = ev.Value
+	}
+
+	pullPolicy := buildah.PullIfMissing
+	if opts.Pull {
+		glog.V(0).Infof("Forcing fresh pull of base image.")
+		pullPolicy = buildah.PullAlways
+	}
+
+	squash := false
+	layers := false
+	switch optimization {
+	case buildapiv1.ImageOptimizationDaemonless:
+		// layers: false, squash: false
+	case buildapiv1.ImageOptimizationDaemonlessWithLayers:
+		layers = true
+	case buildapiv1.ImageOptimizationDaemonlessSquashed:
+		squash = true
+	default:
+		return fmt.Errorf("internal error: image optimization policy %q not fully implemented", string(optimization))
+	}
+
+	var systemContext types.SystemContext
+	if credsDir, ok := os.LookupEnv("PULL_DOCKERCFG_PATH"); ok {
+		systemContext.AuthFilePath = filepath.Join(credsDir, "config.json")
+	}
+	for registry, ac := range opts.AuthConfigs.Configs {
+		glog.V(0).Infof("Setting authentication for registry %q at %q.", registry, ac.ServerAddress)
+		if err := config.SetAuthentication(&systemContext, registry, ac.Username, ac.Password); err != nil {
+			return err
+		}
+		if err := config.SetAuthentication(&systemContext, ac.ServerAddress, ac.Username, ac.Password); err != nil {
+			return err
+		}
+	}
+
+	var mounts []imagebuildah.Mount
+	if opts.BuildBinds != "" {
+		var binds []string
+		err := json.Unmarshal([]byte(opts.BuildBinds), &binds)
+		if err != nil {
+			glog.V(0).Infof("Error parsing binds %q: %v", opts.BuildBinds, err)
+		} else {
+			for _, bind := range binds {
+				spec := strings.SplitN(bind, ":", 2)
+				if len(spec) == 2 && spec[0] != "" && spec[1] != "" {
+					glog.V(0).Infof("Binding %q on host to %q.", spec[0], spec[1])
+					mounts = append(mounts, imagebuildah.Mount{
+						Source:      spec[0],
+						Destination: spec[1],
+						Type:        "bind",
+						Options:     []string{"bind", "private", "nodev", "nosuid"},
+					})
+				}
+			}
+		}
+	}
+
+	options := imagebuildah.BuildOptions{
+		ContextDirectory: dir,
+		PullPolicy:       pullPolicy,
+		TransientMounts:  mounts,
+		Args:             args,
+		Output:           opts.Name,
+		Out:              opts.OutputStream,
+		Err:              opts.OutputStream,
+		ReportWriter:     opts.OutputStream,
+		OutputFormat:     imagebuildah.Dockerv2ImageFormat,
+		SystemContext:    &systemContext,
+		NamespaceOptions: buildah.NamespaceOptions{
+			{Name: string(specs.NetworkNamespace), Host: true},
+		},
+		CommonBuildOpts: &buildah.CommonBuildOptions{
+			Memory:       opts.Memory,
+			MemorySwap:   opts.Memswap,
+			CgroupParent: opts.CgroupParent,
+		},
+		Squash:                  squash,
+		Layers:                  layers,
+		NoCache:                 opts.NoCache,
+		RemoveIntermediateCtrs:  opts.RmTmpContainer,
+		ForceRmIntermediateCtrs: opts.ForceRmTmpContainer,
```

Comment:
will this remove the final container, in the event that the build is unsuccessful?

Reply:
That should happen regardless of these settings, unless I'm missing something.

Reply:
@umohnani8 has clued me in that the default is to leave the final container there for troubleshooting and debugging, so the next revision will set `ForceRmIntermediateCtrs` to `true`.